payment/controllers.go

```diff
@@ -704,3 +705,102 @@ func VerifyCredential(service *Service) handlers.AppHandler {
+		}
+
+		event, err := webhook.ConstructEvent(
+			b, r.Header.Get("Stripe-Signature"), endpointSecret)
+		if err != nil {
+			return handlers.WrapError(err, "error verifying webhook signature", http.StatusBadRequest)
+		}
+
+		// Handle invoice events
+		if event.Type == StripePaymentUpdated {
+			// Retrieve invoice from update events
+			var invoice stripe.Invoice
+			err := json.Unmarshal(event.Data.Raw, &invoice)
+			if err != nil {
+				return handlers.WrapError(err, "error parsing webhook JSON", http.StatusBadRequest)
+			}
+
+			subscription, err := service.scClient.Subscriptions.Get(invoice.Subscription.ID, nil)
+			if err != nil {
+				return handlers.WrapError(err, "error retrieving subscription", http.StatusInternalServerError)
+			}
+			orderID, err := uuid.FromString(subscription.Metadata["orderID"])
+			if err != nil {
+				return handlers.WrapError(err, "error retrieving orderID", http.StatusInternalServerError)
+			}
+
+			// If the invoice is paid set order status to paid, otherwise
+			if invoice.Paid {
+				err = service.Datastore.UpdateOrder(orderID, "paid")
+				if err != nil {
+					return handlers.WrapError(err, "error updating order status", http.StatusInternalServerError)
+				}
+				err = service.Datastore.UpdateOrderMetadata(orderID, "stripeSubscriptionId", subscription.ID)
+				if err != nil {
+					return handlers.WrapError(err, "error updating order metadata", http.StatusInternalServerError)
+				}
+				return handlers.RenderContent(r.Context(), "payment successful", w, http.StatusOK)
+			}
+			err = service.Datastore.UpdateOrder(orderID, "pending")
+			if err != nil {
+				return handlers.WrapError(err, "error updating order status", http.StatusInternalServerError)
+			}
+			err = service.Datastore.UpdateOrderMetadata(orderID, "stripeSubscriptionId", subscription.ID)
+			if err != nil {
+				return handlers.WrapError(err, "error updating order metadata", http.StatusInternalServerError)
+			}
+			return handlers.RenderContent(r.Context(), "payment failed", w, http.StatusOK)
+		}
+
+		// Handle subscription cancellations
+		if event.Type == StripeCustomerSubscriptionDeleted {
+			var subscription stripe.Subscription
+			err := json.Unmarshal(event.Data.Raw, &subscription)
+			if err != nil {
+				return handlers.WrapError(err, "error parsing webhook JSON", http.StatusBadRequest)
+			}
+			orderID, err := uuid.FromString(subscription.Metadata["orderID"])
+			if err != nil {
+				return handlers.WrapError(err, "error retrieving orderID", http.StatusInternalServerError)
+			}
+			err = service.Datastore.UpdateOrder(orderID, "canceled")
+			if err != nil {
+				return handlers.WrapError(err, "error updating order status", http.StatusInternalServerError)
+			}
+			return handlers.RenderContent(r.Context(), "subscription canceled", w, http.StatusOK)
+		}
+
+		return handlers.RenderContent(r.Context(), "event received", w, http.StatusOK)
```

Comment:
This seems like we might be dropping something useful if something does change. Is it possible to log somewhere any abstract message and we can perhaps revisit it later?

Reply:
:+1:

Reply:
resolved: https://github.com/brave-intl/bat-go/pull/816/commits/6276de6cc2ffc85e725c37c13ded51bd438902a1